File: src/optic/lua/flat/ops/FunctionLiteral.java
package optic.lua.flat.ops;

import optic.lua.flat.*;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.stream.Stream;

public class FunctionLiteral implements Step {
	private final StepType type;
	private final List<Step> body;
	private final Register assignTo;
	private final ParameterList params;

	public FunctionLiteral(StepType type, List<Step> body, Register assignTo, ParameterList params) {
		this.type = type;
		this.body = List.copyOf(body);
		this.assignTo = assignTo;
		this.params = params;
	}

	@NotNull
	@Override
	public StepType getType() {
		return type;
	}

	@Override
	public String toString() {
		return typeName() + " " + assignTo + " = function(" + params + ")";
	}

	@Override
	public @NotNull Stream<Step> children() {
		return body.stream();
	}
}

File: src/optic/lua/asm/instructions/FunctionLiteral.kt
package optic.lua.asm.instructions

import optic.lua.asm.AsmBlock
import optic.lua.asm.ParameterList
import optic.lua.asm.Register
import optic.lua.asm.Step

@Deprecated("Use RValue.FunctionLiteral in future")
class FunctionLiteral(val body: AsmBlock, val assignTo: Register, val params: ParameterList) : Step {
    override fun toString(): String {
        return "function $assignTo = function($params)"
    }

    override fun children(): List<Step> = body.steps()

    override fun modified(): Register? = assignTo
}
